query/src/main/java/com/connexta/replication/query/QueryServiceTools.java

```diff
@@ -0,0 +1,146 @@
+
+  /**
+   * Creates a new instance of QueryServiceTools.
+   *
+   * @param nodeAdapterFactories An service which provides a way to retrieve {@link
+   *     com.connexta.replication.api.NodeAdapterFactory}s
+   * @param siteManager A service used to access site configurations
+   * @param filterManager A service used to access filter configurations
+   * @param filterIndexManager A service used to access indexes associated with filter
+   *     configurations
+   * @param queueBroker A service used to retrieve Queues
+   * @param period The length between polling intervals in seconds.
+   */
+  public QueryServiceTools(
+      NodeAdapters nodeAdapterFactories,
+      SiteManager siteManager,
+      FilterManager filterManager,
+      FilterIndexManager filterIndexManager,
+      QueueBroker queueBroker,
+      long period) {
+    this.nodeAdapterFactories = nodeAdapterFactories;
+    this.siteManager = siteManager;
+    this.filterManager = filterManager;
+    this.filterIndexManager = filterIndexManager;
+    this.queueBroker = queueBroker;
+    this.period = period > 0L ? period : DEFAULT_REPLICATION_PERIOD;
+  }
+
+  /**
+   * Gets the amount of time between polling intervals in seconds.
+   *
+   * @return the amount of time between polling intervals in seconds
+   */
+  public long getPeriod() {
+    return period;
+  }
+
+  /**
+   * Returns all the currently configured sites in a {@link Stream}.
+   *
+   * @return All the currently configured sites in a {@link Stream}
+   */
+  public Stream<Site> allSites() {
+    return siteManager.objects();
+  }
+
+  /**
+   * Creates a {@link NodeAdapter} for the given site.
+   *
+   * @param site the site to retrieve a {@link NodeAdapter} for
+   * @return the {@link NodeAdapter} for the given site
+   */
+  public NodeAdapter getAdapterFor(Site site) {
+    return nodeAdapterFactories.factoryFor(site.getType()).create(site.getUrl());
+  }
+
+  /**
+   * Gets all the filters associated with the given site that aren't suspended.
+   *
+   * @param site the site to retrieve all the associated filters for
+   * @return A {@link Stream} containing all the non-suspended filters for the associated site
+   */
+  public Stream<Filter> activeFiltersFor(Site site) {
+    return filterManager.filtersForSite(site.getId()).filter(f -> !f.isSuspended());
+  }
+
+  /**
+   * Gets the {@link FilterIndex} associated with the given filter. If there is no associated filter
+   * index, then a new one is created.
+   *
+   * @param filter the filter to get or create a filter index for
+   * @return the filter index for the given filter
+   */
+  public FilterIndex getOrCreateFilterIndex(Filter filter) {
+    return filterIndexManager.getOrCreate(filter);
+  }
+
+  /**
+   * Persists the given filter index.
+   *
+   * @param filterIndex the filter index to persist
+   */
+  public void saveFilterIndex(FilterIndex filterIndex) {
+    filterIndexManager.save(filterIndex);
```

Comment:
Good point. All these methods should document that these exceptions can be thrown and based on following the code, nothing is handling these exceptions which means it will abort everything. We need to add catches in various places where we are calling these persistence method to handle these exceptions (i.e. `ReplicationException`).